### ISSUE TITLE
Refactor provider

### ISSUE THREAD

**Issue by raphaelrobert** [openmls/openmls#1389]
- [ ] Rename `OpenMlsCryptoProvider` to `OpenMlsProvider` and instances like `backend` to `provider`.
 - [ ] Ensure functions only require the parts of the provider they need, e.g. `&impl OpenMlsCrypto` instead of `&impl OpenMlsProvider`.

**Comment by jbis9051** [comment]
Feel free to assign me. I'll try to get to this soon if I find some time.

**Comment by jbis9051** [comment]
It seems very reasonable to me to get rid of `OpenMlsCryptoProvider` (renamed to `OpenMlsProvider`) entirely

**Comment by jbis9051** [comment]
> For private and pub(crate) functions that seems reasonable.

I was getting some type issues. I think we can deal with it in a future PR.

**Comment by jbis9051** [comment]
re: OpenMlsProvider

I understand your concern. Instead of having `.key_store` `.crypto` and `.rand`, could we simply add the bounds directly. For example, a function that needs to use the `OpenMlsKeyStore` and `OpenMlsRand` would have:

```rs
fn foo(backend: &(impl OpenMlsKeyStore + OpenMlsRand) {
  ...
}
```

The benefit of this approach is (a) we only require the minimal functionality actually needed and (b) we clearly indicate what each function does. What frustrates me is that when I pass a function a `OpenMlsProvider` I have no idea if the function is going to use the key store or not. Documentation helps this problem, but expressing it in the type system is infinitely better.

**Comment by jbis9051** [comment]
> The downside is that the traits could now potentially be in conflict when they identical function names. This approach "flattens" the namespace and it's not clear in the library what trait is meant when you call backend.xxx(...).

Rust specifically prevents ambiguous function calls and provides an extremely easy way to resolve them. See here:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bd0379d93373ee32912f9c1e783e50e6

and

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2534b1516a76f0d3fcc285a284d0c74e

> In reality, application will then probably still only have one struct and implement the different traits on that one struct.

I agree, but I still think the clearer traits allow more flexibility and express things better in the type system.

**Comment by raphaelrobert** [comment]
Sure, Rust has your back of course. I just meant that it might be slightly more confusing for OpenMLS developers.

I think we can try this, but I'll defer to @franziskuskiefer for final approval since he wrote the traits.

**Comment by franziskuskiefer** [comment]
> Rust specifically prevents ambiguous function calls and provides an extremely easy way to resolve them

The ambiguity in itself is not a problem, but you have to add extra syntax to make it work, which isn't great.

> I still think using this method allows more flexibility

Can you elaborate on what flexibility you have in mind here?

**Comment by jbis9051** [comment]
> The ambiguity in itself is not a problem, but you have to add extra syntax to make it work, which isn't great.

I don't consider this a major problem.

> Can you elaborate on what flexibility you have in mind here?

Given some recent unreleased changes, I don't think it matters much any more. Never mind.

**Comment by raphaelrobert** [comment]
Closed via #1400.
